sharedbuffers/mapped_struct.py

```diff
@@ -1718,6 +1741,519 @@ def __str__(self):
+                elif not self.bitrep_hi >> i:
+                    break
+            return out
+
+        dcode, objlen, itemsize, offset, _struct = self.objlist._metadata()
+        for i in xrange(len(self)):
+            out.add(self.objlist._c_getitem(i, dcode, objlen, itemsize, offset, _struct, None))
+        return out
+
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _union_2(self, seq):
+        if not seq:
+            return self
+        elif self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                return proxied_frozenset(
+                    None, self.bitrep_lo | pfset.bitrep_lo, self.bitrep_hi | pfset.bitrep_hi)
+
+        rv = self._add_to(set())
+        rv.update(seq)
+        return frozenset(rv)
+
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _intersect_2(self, seq):
+        if self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                return proxied_frozenset(
+                    None, self.bitrep_lo & pfset.bitrep_lo, self.bitrep_hi & pfset.bitrep_hi)
+        elif not seq:
+            return proxied_frozenset(None, 0, 0)
+
+        rv = self._add_to(set())
+        rv.intersection_update(seq)
+        return frozenset(rv)
+
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _diff_2(self, seq):
+        if not seq:
+            return self
+        elif self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                return proxied_frozenset(
+                    None, self.bitrep_lo & ~pfset.bitrep_lo, self.bitrep_hi & ~pfset.bitrep_hi)
+
+        rv = self._add_to(set())
+        rv.difference_update(seq)
+        return frozenset(rv)
+
+    @cython.ccall
+    @cython.locals(pfset='proxied_frozenset')
+    def _symdiff_2(self, seq):
+        if not seq:
+            return self
+        elif self.objlist is None and type(seq) is proxied_frozenset:
+            pfset = cython.cast(proxied_frozenset, seq)
+            if pfset.objlist is None:
+                return proxied_frozenset(
+                    None, self.bitrep_lo ^ pfset.bitrep_lo, self.bitrep_hi ^ pfset.bitrep_hi)
+
+        rv = self._add_to(set())
+        rv.symmetric_difference_update(seq)
+        return frozenset(rv)
+
+    def union(self, *seqs):
+        if not seqs:
+            return self
+        elif len(seqs) == 1:
+            return self._union_2(seqs[0])
+        return frozenset(self).union(*seqs)
+
+    def intersection(self, *seqs):
+        if not seqs:
+            return self
+        elif len(seqs) == 1:
+            return self._intersect_2(seqs[0])
+        return frozenset(self).intersection(*seqs)
```

Comment:
Again, you ought to implement incremental intersection with intersect_2, or you won't reap the benefits of any optimization in intersect_2.

Same with all the other methods.

Many of them don't even require using the _2 version as a base. Difference can be implemented just as difference_2 only you check for membership against multiple sequences rather than just one.

Try to have good implementations for all the operations.

Reply:
I've done some measuring already, and I focused on making the performance of the 2 argument functions because from what I could tell, once you get into variadic functions, it's a lost battle because of CPython's calling convention.